python_bindings/correctness/pystub.py

```diff
@@ -14,172 +17,165 @@ def _realize_and_check(f, offset = 0):
+        f = cls.call(ctx, buffer_input=b_in, float_arg=3.5, offset=k, nonexistent_generator_param="wat")
     except RuntimeError as e:
-        assert "has no GeneratorParam named: nonexistent_generator_param" in str(e)
+        assert "has no GeneratorParam named: nonexistent_generator_param" in str(e) or \
+               "has no GeneratorParam(s) named: ['nonexistent_generator_param']" in str(e)
     else:
         assert False, 'Did not see expected exception!'
 
-def test_looplevel(gen):
-    x, y = hl.Var('x'), hl.Var('y')
-    target = hl.get_jit_target_from_environment()
-
-    buffer_input = hl.Buffer(hl.UInt(8), [4, 4])
-    buffer_input.fill(123)
-
-    func_input = hl.Func("func_input")
-    func_input[x, y] = x + y
-
-    simple_compute_at = hl.LoopLevel()
-    simple = gen(target, buffer_input, func_input, 3.5,
-        compute_level=simple_compute_at)
-
-    computed_output = hl.Func('computed_output')
-    computed_output[x, y] = simple[x, y] + 3
-
-    simple_compute_at.set(hl.LoopLevel(computed_output, x))
-
-    _realize_and_check(computed_output, 3)
-
-
-def _make_constant_image():
-    constant_image = hl.Buffer(hl.UInt(8), [32, 32, 3], 'constant_image')
+def _make_constant_image(type):
+    constant_image = hl.Buffer(type, [32, 32, 3], 'constant_image')
     for x in range(32):
         for y in range(32):
             for c in range(3):
                 constant_image[x, y, c] = x + y + c
     return constant_image
 
-def test_complex(gen):
-    constant_image = _make_constant_image()
+def test_complex(cls):
+    constant_image = _make_constant_image(hl.UInt(8))
+    constant_image_u16 = _make_constant_image(hl.UInt(16))
     input = hl.ImageParam(hl.UInt(8), 3, 'input')
     input.set(constant_image)
 
     x, y, c = hl.Var(), hl.Var(), hl.Var()
     target = hl.get_jit_target_from_environment()
+    ctx = hl.GeneratorContext(target)
 
     float_arg = 1.25
     int_arg = 33
 
-    func_input = hl.Func("func_input")
-    func_input[x, y, c] = hl.u16(x + y + c)
-
-    r = gen(target,
+    r = cls.call(ctx,
             typed_buffer_input=constant_image,
             untyped_buffer_input=constant_image,
-            simple_input=input,
-            array_input=[ input, input ],
+            simple_input=constant_image,
             float_arg=float_arg,
-            int_arg=[ int_arg, int_arg ],
-            untyped_buffer_output_type="uint8",
-            extra_func_input=func_input,
-            vectorize=True)
+            int_arg=int_arg,
+            extra_input=constant_image_u16,
+            vectorize=True,
+            # We can put GeneratorParams anywhere in the list we want --
+            # they will be examined and applied before anything else --
+            # but it's usually better form to put them all at the end.
+            #
+            # We can specify a halide Type via string or object here
+            simple_input__type=hl.UInt(8),
```

Comment:
Note, the stuff regarding "__" here is out of date and will get updated after https://github.com/halide/Halide/pull/6742 lands (assuming it gets full approval)